File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/typeinfo/FieldPropertyInfo.java
package de.osthus.ambeth.typeinfo;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.objectcollector.IThreadLocalObjectCollector;

public class FieldPropertyInfo extends AbstractPropertyInfo
{
	protected boolean writable;

	public FieldPropertyInfo(Class<?> entityType, String propertyName, Field field)
	{
		this(entityType, propertyName, field, null);
	}

	public FieldPropertyInfo(Class<?> entityType, String propertyName, Field field, IThreadLocalObjectCollector objectCollector)
	{
		super(entityType, objectCollector);
		field.setAccessible(true);
		backingField = field;
		name = propertyName;
		declaringType = field.getDeclaringClass();
		propertyType = field.getType();
		elementType = TypeInfoItemUtil.getElementTypeUsingReflection(propertyType, field.getGenericType());
		init(objectCollector);
	}

	@Override
	protected void init(IThreadLocalObjectCollector objectCollector)
	{
		putAnnotations(backingField);
		super.init(objectCollector);

		writable = (Modifier.isPublic(backingField.getModifiers()) || Modifier.isProtected(backingField.getModifiers()))
				&& !Modifier.isFinal(backingField.getModifiers());
	}

	@Override
	public boolean isReadable()
	{
		return true;
	}

	@Override
	public boolean isWritable()
	{
		return writable;
	}

	@Override
	public boolean isFieldWritable()
	{
		return writable;
	}

	@Override
	public Object getValue(Object obj)
	{
		try
		{
			return backingField.get(obj);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	@Override
	public void setValue(Object obj, Object value)
	{
		try
		{
			backingField.set(obj, value);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}
}

File: ambeth/Ambeth.Util/ambeth/typeinfo/FieldPropertyInfo.cs
using System;
using System.Reflection;

namespace De.Osthus.Ambeth.Typeinfo
{
    public class FieldPropertyInfo : AbstractPropertyInfo
    {
        public FieldPropertyInfo(Type entityType, String propertyName, FieldInfo field)
            : base(entityType)
        {
            EntityType = entityType;
            BackingField = field;
			AddModifiers(field);
            Name = propertyName;
            DeclaringType = field.DeclaringType;
            PropertyType = field.FieldType;
            ElementType = TypeInfoItemUtil.GetElementTypeUsingReflection(PropertyType, null);

            IsWritable = field.IsPublic || field.IsFamily;
            IsReadable = IsWritable;

            Init();
        }

        protected override void Init()
        {
            PutAnnotations(BackingField);

            base.Init();
        }

        public override Object GetValue(Object obj)
        {
            return BackingField.GetValue(obj);
        }

        public override void SetValue(Object obj, Object value)
        {
            BackingField.SetValue(obj, value);
        }
    }
}
